src/sender.js

```diff
@@ -431,6 +596,115 @@ function startNewRow(sender) {
     sender.hasColumns = false;
 }
 
+function createRequestOptions(sender, data) {
+    const timeoutMillis = (data.length / sender.requestMinThroughput) * 1000 + sender.requestTimeout;
+    const options = {
+        hostname: sender.host,
+        port: sender.port,
+        path: '/write?precision=n',
+        method: 'POST',
+        timeout: timeoutMillis
+    };
+    if (sender.secure) {
+        options.rejectUnauthorized = sender.tlsVerify;
+        options.ca = sender.tlsCA;
+    }
+    return options;
+}
+
+function sendHttp(sender, request, options, data, retryTimeout, retryBegin = -1, retryInterval = -1) {
+    return new Promise((resolve, reject) => {
+        const req = request(options, async response => {
+            const body = [];
+            response
+                .on('data', chunk => {
+                    body.push(chunk);
+                })
+                .on('error', err => {
+                    console.error("resp err=" + err);
+                });
+
+            if (response.statusCode === HTTP_NO_CONTENT) {
+                response.on('end', () => {
+                    if (body.length > 0) {
+                        sender.log('warn', `Unexpected message from server: ${Buffer.concat(body)}`);
+                    }
+                    sender.doResolve(resolve);
+                });
+            } else {
+                if (isRetryable(response.statusCode) && retryTimeout > 0) {
+                    if (retryBegin < 0) {
+                        retryBegin = Date.now();
+                        retryInterval = 10;
+                    } else {
+                        const elapsed = Date.now() - retryBegin;
+                        if (elapsed > retryTimeout) {
+                            reject(new Error(`HTTP request failed, statusCode=${response.statusCode}, error=${Buffer.concat(body)}`));
+                            return;
+                        }
+                    }
+                    const jitter = Math.floor(Math.random() * 10) - 5;
+                    await sleep(retryInterval + jitter);
```

Comment:
What would happen if the server/network is slow and we receive `res.on('end')` while waiting for this `sleep` function call? In this case, `sender.doResolve` will be called and a bit later we'll send one more request with `await sendHttp()`.

A safer approach would be to use `setTimeout` and store the timer id, so that we can cancel the timer if the original request succeeds before the retry timeout. On the other hand, we should cancel the original request when we retry (with `req.destroy()` or `req.abort()`).

Reply:
This runs only when we know that request failed with a retryable error.
This includes timeouts, that is true, but there is callback registered for the timeout event which calls `req.destroy()`:
```
req.on('timeout', () => req.destroy(new Error('HTTP request timeout')));
```
I think this will be called.

This `sleep()` is really just to delay the resend after the request has already failed.
I should probably add a test where the request fails with timeout, and assert that `req.destroy()` is called.

Reply:
Right, we probably don't need to cancel the timer. But at this point, we don't know if the response body was fully read, i.e. `response.on('data')/'end'/'error'` may be still called. So, the suggestion is to cancel the request before sleeping and retrying it.

Reply:
We could call `req.destroy()` with a custom error object - in that case that object will be provided to the error listener. This way we can differentiate cancellation from network errors.

Reply:
Yes, that is exactly what I did.
See my comment above, and commit from yesterday:
7cb6c57a1762bb2af77252b0d35883de7bb1bf72